src/util/KinBodyParser.cpp

```diff
@@ -0,0 +1,532 @@
+  // * If <Geom>'s render attribute is false, display nothing. Otherwise...
+  // * If <Render> is present, render the trimesh at the specified path.
+  //   Otherwise...
+  // * Render the collision geometry.
+  //
+
+  // `render` is an optional attribute and true by default.
+  bool renderAttr = true;
+  if (dart::utils::hasAttribute(geomEle, "render"))
+    renderAttr = dart::utils::getAttributeBool(geomEle, "render");
+
+  if (!renderAttr)
+    return;
+
+  const auto hasRenderEle = dart::utils::hasElement(geomEle, "render");
+
+  if (hasRenderEle)
+  {
+    auto fileNameAndScaleOfRender = resolveFileNameAndScale(geomEle, "render");
+    auto fileNameOfRender = fileNameAndScaleOfRender.first;
+    auto uniScaleOfRender = fileNameAndScaleOfRender.second;
+    auto renderShapeNodeName = resolveShapeName(geomEle, bodyNode);
+
+    if (typeAttr == "trimesh"
+        && fileNameOfCollision == fileNameOfRender
+        && uniScaleOfCollision == uniScaleOfRender)
+    {
+      shapeNode->createVisualAspect();
+    }
+    else
+    {
+      auto shape = readMeshShape(
+          fileNameOfRender, uniScaleOfRender, baseUri, retriever);
+
+      // Create additional ShapeNode only for visualization
+      bodyNode->createShapeNodeWith<
+          dart::dynamics::VisualAspect>(shape, renderShapeNodeName);
+    }
+  }
+  else
+  {
+    if (!shapeNode)
+      shapeNode->createVisualAspect();
+    // `none` type geom can reach here that hasn't created a ShapeNode.
+  }
+}
+
+//==============================================================================
+dart::common::ResourceRetrieverPtr getRetriever(
+  const dart::common::ResourceRetrieverPtr& nullOrRetriever)
+{
+  if (nullOrRetriever)
+    return nullOrRetriever;
+  else
+    return std::make_shared<dart::common::LocalResourceRetriever>();
+}
+
+//==============================================================================
+std::string toLowerCases(const std::string& in)
+{
+  std::string out(in);
+  std::transform(out.begin(), out.end(), out.begin(), ::tolower);
+
+  return out;
+}
+
+//==============================================================================
+void transformElementNamesToLowerCasesRecurse(tinyxml2::XMLElement* ele)
+{
+  std::string eleName = ele->Name();
+  ele->SetName(toLowerCases(eleName).c_str());
+
+  auto childEle = ele->FirstChildElement();
+  while (childEle)
+  {
+    transformElementNamesToLowerCasesRecurse(childEle);
+    childEle = childEle->NextSiblingElement();
+  }
+}
+
+//==============================================================================
+tinyxml2::XMLElement* transformElementNamesToLowerCases(
+    tinyxml2::XMLDocument& doc)
+{
+  auto firstEle = doc.FirstChildElement();
+  if (firstEle)
+    transformElementNamesToLowerCasesRecurse(firstEle);
+}
+
+//==============================================================================
+std::vector<std::string> split(
+    const std::string& str, const std::string& delimiters)
```

Comment:
Nice helper function. I would be open to moving this to `util`, since it may be useful elsewhere.